### PR TITLE
Add implementation of NicConfigurationTemplate reconciler

### PR DESCRIPTION
**Pull request by almaslennikov**
This reconciler is getting a list of NicDevice matching the selectors and applies the template spec to them

### CODE REVIEW

File: internal/controller/nicconfigurationtemplate_controller.go [hunk 1]
```diff
@@ -43,26 +57,262 @@ type NicConfigurationTemplateReconciler struct {
...
+		if err != nil {
+			log.Log.Error(err, "failed to update template status", "template", template.Name)
+			return ctrl.Result{}, err
+		}
+	}
 
 	return ctrl.Result{}, nil
 }
 
+func (r *NicConfigurationTemplateReconciler) dropDeviceFromStatus(ctx context.Context, deviceName string, template *v1alpha1.NicConfigurationTemplate) error {
```
Comment (ykulazhenkov) [inline internal/controller/nicconfigurationtemplate_controller.go:174]: error is never used, let's drop it

File: internal/controller/nicconfigurationtemplate_controller.go [hunk 2]
```diff
@@ -43,26 +57,262 @@ type NicConfigurationTemplateReconciler struct {
...
+		if !ok {
+			log.Log.Info("device doesn't match any node, skipping", "device", device.Name)
+			continue
+		}
+
+		var matchingTemplates []*v1alpha1.NicConfigurationTemplate
+
+		for _, template := range templates {
+			if !deviceMatchesSelectors(&device, template, node) {
+				err := r.dropDeviceFromStatus(ctx, device.Name, template)
```
Comment (ykulazhenkov) [inline internal/controller/nicconfigurationtemplate_controller.go:111]: this function never returns errors

File: internal/controller/nicconfigurationtemplate_controller.go [hunk 3]
```diff
@@ -43,26 +57,262 @@ type NicConfigurationTemplateReconciler struct {
...
 
-	// TODO(user): your logic here
+			templateNames := []string{}
+			for _, template := range matchingTemplates {
+				templateNames = append(templateNames, template.Name)
+			}
+			joinedTemplateNames := strings.Join(templateNames, ",")
+			err = fmt.Errorf("device matches several configuration templates: %s, %s", device.Name, joinedTemplateNames)
+			log.Log.Error(err, "Device matches several configuration templates, deleting its spec and emitting an error event", "device", device.Name)
+			return ctrl.Result{}, r.handleErrorSeveralMatchingTemplates(ctx, &device, joinedTemplateNames)
```
Comment (adrianchiris) [inline internal/controller/nicconfigurationtemplate_controller.go:146]: why do we want to exit early and not continue to other devices ?
i would return only if error happened (then we reconcile again)

File: internal/controller/nicconfigurationtemplate_controller.go [hunk 4]
```diff
@@ -43,26 +57,262 @@ type NicConfigurationTemplateReconciler struct {
...
+		if len(matchingTemplates) == 0 {
+			log.Log.V(2).Info("Device doesn't match any configuration template, resetting the spec", "device", device.Name)
+			device.Spec.Configuration = nil
+			err = r.Update(ctx, &device)
+			if err != nil {
+				log.Log.Error(err, "Failed to update device's spec", "device", device)
+				return ctrl.Result{}, err
+			}
+			continue
+		} else if len(matchingTemplates) > 1 {
```
Comment (ykulazhenkov) [inline internal/controller/nicconfigurationtemplate_controller.go:131]: You can move this `else if` condition to an individual `if` case

File: internal/controller/nicconfigurationtemplate_controller.go [hunk 5]
```diff
@@ -43,26 +57,262 @@ type NicConfigurationTemplateReconciler struct {
...
+			for _, template := range matchingTemplates {
+				err := r.dropDeviceFromStatus(ctx, device.Name, template)
+				if err != nil {
```
Comment (ykulazhenkov) [inline internal/controller/nicconfigurationtemplate_controller.go:134]: the function never return errors

File: internal/controller/nicconfigurationtemplate_controller.go (hunk 3 above)
Comment (ykulazhenkov) [inline internal/controller/nicconfigurationtemplate_controller.go:146]: I think in this case the device name will never be removed from the template status. You removed it from the struct, but didn't updated the API.

File: internal/controller/nicconfigurationtemplate_controller.go [hunk 6]
```diff
@@ -43,26 +57,262 @@ type NicConfigurationTemplateReconciler struct {
...
+			log.Log.Error(err, "failed to update template status", "template", template.Name)
+			return ctrl.Result{}, err
+		}
+	}
 
 	return ctrl.Result{}, nil
 }
 
+func (r *NicConfigurationTemplateReconciler) dropDeviceFromStatus(ctx context.Context, deviceName string, template *v1alpha1.NicConfigurationTemplate) error {
+	if slices.Contains(template.Status.NicDevices, deviceName) {
```
Comment (ykulazhenkov) [inline internal/controller/nicconfigurationtemplate_controller.go:175]: nit: you can save call of Contains - Index will return -1 if the element is not found

File: internal/controller/nicconfigurationtemplate_controller_test.go [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+		Expect(k8sClient.Create(ctx, device6)).To(Succeed())
+		device6.Status = v1alpha1.NicDeviceStatus{
+			Node:         validNode.Name,
+			Type:         "ConnectX6",
+			Ports:        []v1alpha1.NicDevicePortSpec{{PCI: "0000:3b:00.0"}},
+			SerialNumber: "serialNumber3",
+		}
+		Expect(k8sClient.Status().Update(ctx, device6)).To(Succeed())
+
+		Expect(k8sClient).To(Not(BeNil()))
```
Comment (ykulazhenkov) [inline internal/controller/nicconfigurationtemplate_controller_test.go:225]: nit: it looks like it is a bit too late to check this here :) You already used the client multiple times

File: internal/controller/nicconfigurationtemplate_controller.go (hunk 3 above)
Comment (almaslennikov) [reply internal/controller/nicconfigurationtemplate_controller.go]: Changed this one to return an error only if the function failed. Templates are updated together at the end of the reconcile loop

File: internal/controller/nicconfigurationtemplate_controller_test.go (hunk 7 above)
Comment (almaslennikov) [reply internal/controller/nicconfigurationtemplate_controller_test.go]: You're right :) Dropped this check, it's not needed actually

File: internal/controller/nicconfigurationtemplate_controller.go (hunk 6 above)
Comment (almaslennikov) [reply internal/controller/nicconfigurationtemplate_controller.go]: Thanks for the tip! Applied it

File: internal/controller/nicconfigurationtemplate_controller.go (hunk 5 above)
Comment (almaslennikov) [reply internal/controller/nicconfigurationtemplate_controller.go]: Yep, thanks, dropped the error handling

File: Makefile.version [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+ifneq (,$(GIT_TAG))
+    RELEASE_STATUS = ",released"
+endif
+
+ifeq (,$(VERSION))
+    VERSION = $(GIT_TAG_LAST)
+endif
+
+# Add version/commit/date to linker flags
+VERSION_LDFLAGS = "-X github.com/Mellanox/nic-configuration-operator/internal/version.version=${VERSION} \
```
[the remark this answers is not in the document]
Comment (almaslennikov) [reply Makefile.version:35]: Those aren't files but linker flags to build metadata into the binary

File: Makefile [hunk 9]
```diff
@@ -1,9 +1,12 @@
+# Version information
+include Makefile.version
+
 # VERSION defines the project version for the bundle.
 # Update this value when you upgrade the version of your project.
 # To re-generate a bundle for another specific version without changing the standard setup, you can:
 # - use the VERSION as arg of the bundle target (e.g make bundle VERSION=0.0.2)
 # - use environment variables to overwrite this value (e.g export VERSION=0.0.2)
-VERSION ?= 0.0.1
+# VERSION ?= 0.0.1
```
[the remark this answers is not in the document]
Comment (almaslennikov) [reply Makefile]: Dropped it, thanks

File: internal/controller/nicconfigurationtemplate_controller.go (hunk 3 above)
Comment (almaslennikov) [reply internal/controller/nicconfigurationtemplate_controller.go]: You're right, not it only returns if an error occurred

File: internal/controller/nicconfigurationtemplate_controller.go [hunk 10]
```diff
@@ -43,26 +57,262 @@ type NicConfigurationTemplateReconciler struct {
...
+	nicDeviceEventHandler := handler.Funcs{
+		UpdateFunc: func(ctx context.Context, e event.UpdateEvent, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+			log.Log.Info("Enqueuing sync for update event", "resource", e.ObjectNew.GetName())
+			qHandler(q)
+		},
+	}
+
 	return ctrl.NewControllerManagedBy(mgr).
-		For(&configurationnetv1alpha1.NicConfigurationTemplate{}).
+		For(&v1alpha1.NicConfigurationTemplate{}).
```
[the remark this answers is not in the document]
Comment (almaslennikov) [reply internal/controller/nicconfigurationtemplate_controller.go]: Dropped it, thx

File: Makefile.version (hunk 8 above)
[the remark this answers is not in the document]
Comment (ykulazhenkov) [reply Makefile.version:35]: Yes, but usually these flags are used to set values in existing files, like this https://github.com/Mellanox/maintenance-operator/blob/main/internal/version/version.go

[the remark this answers is not in the document]
Comment (almaslennikov) [reply Makefile.version]: yeah, you're right. Added version handling in the manager binary

**Review by adrianchiris — APPROVED** [review]

**Review by rollandf — APPROVED** [review]

**Pull request merged**